File: java/ingest/DeleteDataset.java
import java.io.*;
import java.util.Properties;

import javax.xml.xpath.*;

import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpDelete;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.DefaultHttpClient;
import org.xml.sax.InputSource;

/**
 * This example shows how to delete a dataset in ECHO using the Catalog REST
 * API.
 * 
 * @author jgilman
 */
public class DeleteDataset
{

  public static void main(String[] args) throws Exception
  {
    // Get username and password
    Properties properties = loadProperties();
    String username = properties.getProperty("username");
    String password = properties.getProperty("password");
    String provider = properties.getProperty("provider");

    // Create the HTTPClient
    DefaultHttpClient httpClient = new DefaultHttpClient();

    String token = getToken(httpClient, username, password);
    deleteDataset(httpClient, token, provider);
  }

  /**
   * Ingests a dataset into Catalog REST.
   * 
   * @param httpClient
   *          the http client to use to send the data
   * @param token
   *          A token with access to ingest data for the provider
   * @param provider
   *          the provider id to ingest data for.
   * @throws Exception
   */
  private static void deleteDataset(HttpClient httpClient, String token,
      String provider) throws Exception
  {
    String datasetId = "SampleValidCollection_10";
    String url = "https://testbed.echo.nasa.gov/catalog-rest/providers/"
        + provider + "/datasets/" + datasetId;
    HttpDelete httpDelete = new HttpDelete(url);

    // Set the Echo-Token Header
    httpDelete.setHeader("Echo-Token", token);

    // Send the request
    HttpResponse response = httpClient.execute(httpDelete);

    // Check for a successful response
    int responseCode = response.getStatusLine().getStatusCode();

    if (responseCode == 204)
    {
      System.out.println("Dataset deleted.");
    }
    else if (responseCode == 404)
    {
      System.out.println("Dataset did not exist");
    }
    else
    {
      System.out.println("Unexpected response code: " + responseCode);
      String body = readInputStream(response.getEntity().getContent());
      System.out.println(" body: " + body);
    }
    httpDelete.releaseConnection();
  }

  /**
   * Logs in and returns a token.
   * 
   * @return the token.
   * @throws Exception
   */
  private static String getToken(HttpClient httpClient, String username,
      String password) throws Exception
  {
    HttpPost httpPost = new HttpPost(
        "https://testbed.echo.nasa.gov/echo-rest/tokens");

    StringEntity requestEntity = new StringEntity(getLoginPostBody(username,
        password));
    requestEntity.setContentType("application/xml");
    httpPost.setEntity(requestEntity);

    HttpResponse response = httpClient.execute(httpPost);
    String token;
    try
    {
      String xmlResponse = readInputStream(response.getEntity().getContent());
      // Parse out the token value from the XML response
      token = applyXPath(xmlResponse, "/token/id");
    }
    finally
    {
      httpPost.releaseConnection();
    }
    return token;
  }

  /**
   * Returns the POST body to login.
   * 
   * @param username
   * @param password
   * @return XML POST body to login
   */
  private static String getLoginPostBody(String username, String password)
  {
    return "<token><username>"
        + username
        + "</username><password>"
        + password
        + "</password><client_id>ETIM Java</client_id><user_ip_address>127.0.0.1</user_ip_address></token>";
  }

  /**
   * Applies an XPath to an XML string and returns the text result.
   * 
   * @param xml
   * @param xpath
   * @return Result of applying XPath.
   * @throws XPathExpressionException
   */
  private static String applyXPath(String xml, String xpath)
      throws XPathExpressionException
  {
    XPath evaluator = XPathFactory.newInstance().newXPath();
    return evaluator.evaluate(xpath, new InputSource(new StringReader(xml)));
  }

  /**
   * Reads an input stream and returns the contents as a string.
   * 
   * @param is
   *          input stream
   * @return string contents of input string
   * @throws IOException
   */
  private static String readInputStream(InputStream is) throws IOException
  {
    LineNumberReader reader = new LineNumberReader(new InputStreamReader(is));
    try
    {
      StringBuilder builder = new StringBuilder();
      String line = null;
      while ((line = reader.readLine()) != null)
      {
        if (builder.length() != 0)
        {
          builder.append("\n");
        }
        builder.append(line);
      }
      return builder.toString();
    }
    finally
    {
      reader.close();
    }
  }

  /**
   * Loads the Ingest config properties.
   * 
   * @return properties containing username, password, and provider.
   * @throws Exception
   */
  private static Properties loadProperties() throws Exception
  {
    FileInputStream in = null;
    Properties properties = new Properties();
    try
    {
      in = new FileInputStream("../../ingest_config.properties");
      properties.load(in);
    }
    finally
    {
      if (in != null)
      {
        in.close();
      }
    }
    return properties;
  }

}

File: perl/ingest/delete_dataset.pl
#!/usr/bin/env perl

use strict;
use warnings;

use Data::Dumper;
use File::Slurp;
use HTTP::Status qw(:constants);
use REST::Client;
use XML::Simple;
use Config::Properties;

sub get_token {
  my ($username, $password) = @_;

  # The REST endpoint to create a token in the ECHO Testbed environment
  my $echo_rest_tokens_url = 'https://testbed.echo.nasa.gov/echo-rest/tokens';

  # Create the REST client
  my $client = REST::Client->new();

  # We'll be using XML here.  We could also have used application/json
  my $request_headers = {
    'Content-Type' => 'application/xml',
    'Accept' => 'application/xml'
  };

  # The token request being sent
  my $token_request_hash_ref = {
    'username' => { 'content' => $username },
    'password' => { 'content' => $password  },
    'client_id' => { 'content' => 'ETIM demo' },
    'user_ip_address' => { 'content' => '127.0.0.1' }
  };
  my $token_request_xml = XMLout($token_request_hash_ref, RootName => 'token');

  # POST the request to the REST API
  $client->POST($echo_rest_tokens_url, $token_request_xml, $request_headers);

  my $token = undef;

  if ($client->responseCode() == HTTP_CREATED) {
    # If we get a 201 status code back, the token was created
    my $response_hash_ref = XMLin($client->responseContent());
    $token = $response_hash_ref->{'id'};
  }
  else {
    # If we get anything other than a 201, something went wrong
    print "Failed to create token\n";
    parse_and_display_errors($client->responseContent());
    exit 1;
  }

  return $token;
}

# Pull out and display the errors
sub parse_and_display_errors {
  my $error_xml = shift;

  my $errors_ref = XMLin($error_xml, ForceArray => qw(error));

  foreach my $error (@{$errors_ref->{'error'}}) {
    my $error_message = $error->{'content'};
    print "  - ${error_message}\n";
  }

  return;
}

sub delete_dataset {
  my ($token, $provider_id, $dataset_id) = @_;

  my $dataset_ingest_url = "https://testbed.echo.nasa.gov/catalog-rest/providers/${provider_id}/datasets/${dataset_id}";

  my $client = REST::Client->new();

  my $request_headers = {
    'Echo-Token' => $token
  };

  # PUT the dataset to the REST API
  $client->DELETE($dataset_ingest_url, $request_headers);

  if ($client->responseCode() == HTTP_NO_CONTENT) {
    # If a 204 status code is returned, the dataset was deleted
    print "Dataset was deleted.\n";
  }
  elsif ($client->responseCode() == HTTP_NOT_FOUND) {
    # If a 404 status code is returned, the dataset did not exist
    print "Dataset did not exist.\n";
  }
  else {
    # If we get anything other than a 200 or 404, something went wrong
    print "Failed to delete dataset\n";
    parse_and_display_errors($client->responseContent());
    exit 1;
  }

  return;
}
open my $fh, '<', '../../ingest_config.properties'
    or die "unable to open configuration file";

my $properties = Config::Properties->new();
$properties->load($fh);

my $provider_id = $properties->getProperty("provider");
my $username = $properties->getProperty("username");
my $password = $properties->getProperty("password");

my $dataset_id = "SampleValidCollection_10";
print "Deleting ${dataset_id}\n";

my $token = get_token($username, $password);
delete_dataset($token, $provider_id, $dataset_id);

